src/anomalib/metrics/per_image/pimo.py

```diff
@@ -0,0 +1,956 @@
+"""Per-Image Overlap curve (PIMO, pronounced pee-mo) and its area under the curve (AUPIMO).
+
+# PIMO
+
+PIMO is a curve of True Positive Rate (TPR) values on each image across multiple anomaly score thresholds.
+The anomaly score thresholds are indexed by a (shared) valued of False Positive Rate (FPR) measure on the normal images.
+
+Each *anomalous* image has its own curve such that the X-axis is shared by all of them.
+
+At a given threshold:
+    X-axis: Shared FPR (may vary)
+        1. Log of the Average of per-image FPR on normal images.
+        SEE NOTE BELOW.
+    Y-axis: per-image TP Rate (TPR), or "Overlap" between the ground truth and the predicted masks.
+
+*** Note about other shared FPR alternatives ***
+The shared FPR metric can be made harder by using the cross-image max (or high-percentile) FPRs instead of the mean.
+Rationale: this will further punish models that have exceptional FPs in normal images.
+So far there is only one shared FPR metric implemented but others will be added in the future.
+
+# AUPIMO
+
+`AUPIMO` is the area under each `PIMO` curve with bounded integration range in terms of shared FPR.
+
+# Disclaimer
+
+This module implements torch interfaces to access the numpy code in `pimo_numpy.py`.
+Tensors are converted to numpy arrays and then passed and validated in the numpy code.
+The results are converted back to tensors and eventually wrapped in an dataclass object.
+
+Validations will preferably happen in ndarray so the numpy code can be reused without torch,
+so often times the Tensor arguments will be converted to ndarray and then validated.
+"""
+
+# Original Code
+# https://github.com/jpcbertoldo/aupimo
+#
+# Modified
+# Copyright (C) 2024 Intel Corporation
+# SPDX-License-Identifier: Apache-2.0
+
+import json
+import logging
+from collections.abc import Sequence
+from dataclasses import dataclass, field
+from pathlib import Path
+
+import torch
+from torch import Tensor
+from torchmetrics import Metric
+
+from anomalib.data.utils.image import duplicate_filename
+from anomalib.data.utils.path import validate_path
+
+from . import _validate, pimo_numpy, utils
+from .binclf_curve_numpy import BinclfAlgorithm
+from .utils import StatsOutliersPolicy, StatsRepeatedPolicy
+
+logger = logging.getLogger(__name__)
+
+# =========================================== AUX ===========================================
+
+
+def _images_classes_from_masks(masks: Tensor) -> Tensor:
+    masks = torch.concat(masks, dim=0)
+    device = masks.device
+    image_classes = pimo_numpy._images_classes_from_masks(masks.numpy())  # noqa: SLF001
+    return torch.from_numpy(image_classes, device=device)
+
+
+# =========================================== ARGS VALIDATION ===========================================
+
+
+def _validate_is_anomaly_maps(anomaly_maps: Tensor) -> None:
```

Comment:
would it be an idea to move these validation utils to _validate.py? these functions already use functions from validate.